crates/re_renderer/shader/rectangle_fs.wgsl

```diff
@@ -80,76 +88,61 @@ fn fs_main(in: VertexOut) -> @location(0) vec4f {
     }
 
     let coord = in.texcoord * texture_dimensions;
-    let clamped_coord = clamp_to_edge_nearest_neighbor(coord, texture_dimensions);
-    let v00_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(-0.5, -0.5), texture_dimensions);
-    let v01_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(-0.5, 0.5), texture_dimensions);
-    let v10_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(0.5, -0.5), texture_dimensions);
-    let v11_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(0.5, 0.5), texture_dimensions);
 
+    let filter_nearest = (tex_filter(coord) == FILTER_NEAREST);
+
+    if filter_nearest {
+        v00_coord = clamp_to_edge_nearest_neighbor(coord, texture_dimensions);
+        v01_coord = v00_coord;
+        v10_coord = v00_coord;
+        v11_coord = v00_coord;
+    } else {
+        v00_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(-0.5, -0.5), texture_dimensions);
+        v01_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(-0.5, 0.5), texture_dimensions);
+        v10_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(0.5, -0.5), texture_dimensions);
+        v11_coord = clamp_to_edge_nearest_neighbor(coord + vec2f(0.5, 0.5), texture_dimensions);
+    }
+
+    // WARNING! WARNING! WARNING! WARNING! WARNING! WARNING! WARNING! WARNING! WARNING! WARNING!
+    // NO MORE SAMPLE TYPES CAN BE ADDED TO THIS SHADER!
+    // The shader is already too large and adding more sample types will push us over the size limit.
+    // See: https://github.com/rerun-io/rerun/issues/3931, https://github.com/rerun-io/rerun/issues/5073
     if rect_info.sample_type == SAMPLE_TYPE_FLOAT {
-        if tex_filter(coord) == FILTER_NEAREST {
-            // nearest
-            normalized_value = decode_color(textureLoad(texture_float,
-                clamp_to_edge_nearest_neighbor(coord, texture_dimensions), 0));
-        } else {
-            // bilinear
-            let v00 = textureLoad(texture_float, v00_coord, 0);
-            let v01 = textureLoad(texture_float, v01_coord, 0);
-            let v10 = textureLoad(texture_float, v10_coord, 0);
-            let v11 = textureLoad(texture_float, v11_coord, 0);
-            normalized_value = decode_color_and_filter_bilinear(coord, v00, v01, v10, v11);
-        }
+        normalized_value = decode_color_and_filter_nearest_or_bilinear(
+            filter_nearest,
+            coord,
+            textureLoad(texture_float, v00_coord, 0),
+            textureLoad(texture_float, v01_coord, 0),
+            textureLoad(texture_float, v10_coord, 0),
+            textureLoad(texture_float, v11_coord, 0));
```

Comment:
This means we're always doing 4 taps, even with nearest neighbor sampling. Not great, not terrible. But worth a comment. I appreciate thought that they all point to the same tex-coord, so should be available in the texture cache.